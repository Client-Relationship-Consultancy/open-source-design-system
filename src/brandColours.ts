export const colourPalette = {
  examplePalette: {
    primary: {
      main: {
        hex: "#A42824",
        RGB: "164,40,36",
        on: "#FFFFFF",
      },
      light: {
        hex: "#DA342F",
        RGB: "218,52,47",
        on: "#FFFFFF",
      },
      dark: {
        hex: "#661614",
        RGB: "102,22,20",
        on: "#FFFFFF",
      },
    },
    secondary: {
      main: {
        hex: "#0B466B",
        RGB: "11,70,107",
        on: "#FFFFFF",
      },
      light: {
        hex: "#599CAF",
        RGB: "89,156,175",
        on: "#FFFFFF",
      },
      dark: {
        hex: "#05263B",
        RGB: "5,38,59",
        on: "#FFFFFF",
      },
    },
    complimentary: {
      main: {
        hex: "#FFEE52",
        RGB: "255,238,82",
        on: "#272727",
      },
      light: {
        hex: "#FFF59D",
        RGB: "255,245,157",
        on: "#272727",
      },
      dark: {
        hex: "#CEB900",
        RGB: "206,185,0",
        on: "#272727",
      },
    },
    action: {
      main: {
        hex: "#743C8F",
        RGB: "116,60,143",
        on: "#FFFFFF",
      },
      light: {
        hex: "#AD65CF",
        RGB: "173,101,207",
        on: "#FFFFFF",
      },
      dark: {
        hex: "#4E2662",
        RGB: "79,38,98",
        on: "#FFFFFF",
      },
    },
    error: {
      main: {
        hex: "#FF7320",
        RGB: "255,115,32",
        on: "#FFFFFF",
      },
      dark: {
        hex: "#C54901",
        RGB: "197,73,1",
        on: "#FFFFFF",
      },
    },
    background: {
      hex: "#F6F6F6",
      RGB: "246,246,246",
    },
    surface: {
      hex: "#FFFFFF",
      RGB: "255,255,255",
      on: "#272727",
    },
    white: {
      hex: "#FFFFFF",
      RGB: "255,255,255",
      on: "#272727",
    },
    black: {
      main: { hex: "#272727", RGB: "39,39,39", on: "#FFFFFF" },
      tint80: { hex: "#696969", RGB: "105,105,105", on: "#FFFFFF" },
      tint60: { hex: "#A0A0A0", RGB: "160,160,160", on: "#FFFFFF" },
      tint40: { hex: "#C7C7C7", RGB: "199,199,199", on: "#272727" },
      tint20: { hex: "#ECECEC", RGB: "236,236,236", on: "#272727" },
    },
  },
};

interface IColourCodes {
  hex: string;
  RGB: string;
  on?: string;
}
interface IColourVariations {
  main: IColourCodes;
  light?: IColourCodes;
  dark?: IColourCodes;
}

interface IBlackVariations {
  main: IColourCodes;
  tint80: IColourCodes;
  tint60: IColourCodes;
  tint40: IColourCodes;
  tint20: IColourCodes;
}

export interface IColourPalette {
  primary: IColourVariations;
  secondary: IColourVariations;
  complimentary: IColourVariations;
  action: IColourVariations;
  error: IColourVariations;
  background: IColourCodes;
  surface: IColourCodes;
  white: IColourCodes;
  black: IBlackVariations;
}
