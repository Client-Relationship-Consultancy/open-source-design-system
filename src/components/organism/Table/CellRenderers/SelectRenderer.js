import React from "react"
import PropTypes from "prop-types"

// Using react-select, read more here. https://github.com/JedWatson/react-select
import ReactSelect from "react-select"
import styles from "../../../atoms/Select/SelectStyles"

export class Select extends React.PureComponent {
  state = {
    menuOpen: false,
  }

  componentDidMount() {
    // this is needed due to a bug of react-select that doesn't consider menuPlacement before is fully rendered
    // here is the issue https://github.com/JedWatson/react-select/issues/3421
    // the issue is closed and claims to be resolved with newer version but ElioS tryed to update it and still experienced the problem
    setTimeout(() => {
      this.setState({ menuOpen: true })
    }, 1)
  }

  renderOptions = () =>
    this.props.values.map((value) => {
      return typeof value === "string"
        ? {
            value,
            label: value,
          }
        : value
    })

  renderValue = () => {
    if (!this.props.value) return null

    return this.renderOptions().find((option) => option.value && option.value === this.props.value)
  }

  render = () => {
    return (
      <ReactSelect
        id="SelectEditor"
        styles={styles(this.props.theme)}
        menuPlacement={this.props.direction}
        value={this.renderValue()}
        options={this.renderOptions()}
        onChange={this.props.onChange}
        isDisabled={false}
        isClearable={this.props.isClearable}
        isMulti={false}
        onBlur={() => {
          this.setState({ menuOpen: false })
        }}
        menuIsOpen={this.state.menuOpen}
      />
    )
  }
}

export default class SelectRenderer extends React.Component {
  state = {
    value: this.props.value,
  }

  handleChange = (selectedOption) => {
    this.setState({ value: selectedOption ? selectedOption.value : selectedOption }, () =>
      this.props.api.stopEditing(false),
    )
  }

  getValue = () => this.state.value

  calculateDirection = () => {
    // get the distance from the top of the body without considering scroll
    const distanceFromTop = this.props.node.rowTop
    // get the body html element
    const tableBodyElement = this.props.agGridReact.eGridDiv.querySelector(".ag-body-viewport")
    // get the ammount scrolled inside the table
    const scrolledDistanceFromTop = tableBodyElement.scrollTop
    // get the height of the table without considering the content
    const tableBodyHeight = tableBodyElement.getBoundingClientRect().height
    // calculate the distance of the row from the top of the table considering the scroll
    const RelativeDistance = distanceFromTop - scrolledDistanceFromTop
    // check if the row is closer to the bottom
    const isRowNearBottom = RelativeDistance > tableBodyHeight / 2
    return isRowNearBottom ? "top" : "bottom"
  }

  render() {
    return (
      <Select
        direction={this.calculateDirection()}
        value={this.state.value}
        values={this.props.values}
        isDisabled={this.props.isDisabled}
        onChange={this.handleChange}
        theme={this.props.theme}
        isClearable={this.props.isClearable}
      />
    )
  }
}

SelectRenderer.propTypes = {
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  values: PropTypes.array.isRequired,
  isDisabled: PropTypes.bool,
  isClearable: PropTypes.bool,
  theme: PropTypes.object,
  node: PropTypes.shape({
    rowTop: PropTypes.number,
  }),
}

Select.propTypes = {
  ...SelectRenderer.propTypes,
  values: PropTypes.array.isRequired,
}

SelectRenderer.displayName = "SelectRenderer"
Select.displayName = "Select"
