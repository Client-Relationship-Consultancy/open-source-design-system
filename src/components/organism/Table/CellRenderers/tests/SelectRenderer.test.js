import React from "react"
import { shallow } from "enzyme"
import "jest-styled-components"
import { colourPalette } from "../../../../../brandColours.ts"
import SelectRenderer, { Select } from "../SelectRenderer"

const theme = {
  props: {
    theme: colourPalette.examplePalette,
  },
}
const api = {
  stopEditing: jest.fn(),
}

const commonOptions = [
  { label: "Option 1", value: "option-1" },
  { label: "Option 2", value: "option-2" },
  { label: "Option 3", value: "option-3" },
]

describe("Test handle Change", () => {
  let component
  beforeEach(() => {
    component = shallow(
      <SelectRenderer
        agGridReact={theme}
        values={["option 1", "option 2"]}
        value="option 1"
        isClearable
        node={{ fromTop: 140 }}
        api={api}
      />,
    )
  })

  it("Sets the selected value on handleChange", () => {
    const newOption = { value: "option value", label: "option label" }
    component.instance().handleChange(newOption)
    expect(component.state().value).toEqual("option value")
    expect(api.stopEditing).toBeCalledWith(false)
  })

  it("Sets the selected value as null when the value is cleared", () => {
    const newOption = null
    component.instance().handleChange(newOption)
    expect(component.state().value).toEqual(newOption)
    expect(api.stopEditing).toBeCalledWith(false)
  })

  it("Test get value returns correct value", () => {
    expect(component.instance().getValue()).toEqual("option 1")
  })

  it("match last snapshot", () => {
    expect(component).toMatchSnapshot()
  })

  it("should handle a falsy value for value prop", () => {
    const tree = shallow(
      <SelectRenderer
        agGridReact={theme}
        values={["option 1", "option 2"]}
        value={null}
        isClearable
        node={{ fromTop: 140 }}
      />,
    )
    const select = tree
      .dive()
      .dive()
      .find("Select")
    expect(select.props().value).toBeFalsy()
  })
})

describe("Select", () => {
  it("match last snapshot", () => {
    const component = shallow(
      <Select agGridReact={theme} isClearable values={["option 1", "option 2"]} value="option 1" />,
    )
    expect(component).toMatchSnapshot()
  })

  it("should render an array of strings into an object with value and label keys", () => {
    const options = ["away", "boom", "caw"]
    const component = shallow(
      <Select agGridReact={theme} isClearable values={options} value="option 1" />,
    )
    expect(component.instance().renderOptions()).toMatchObject([
      { value: "away", label: "away" },
      { value: "boom", label: "boom" },
      { value: "caw", label: "caw" },
    ])
  })

  it("should render the values as is if options is not an array of strings", () => {
    const component = shallow(
      <Select agGridReact={theme} isClearable values={commonOptions} value="option 1" />,
    )
    expect(component.instance().renderOptions()).toMatchObject([
      { label: "Option 1", value: "option-1" },
      { label: "Option 2", value: "option-2" },
      { label: "Option 3", value: "option-3" },
    ])
  })

  it("should render the matching object from the options array of objects", () => {
    const component = shallow(
      <Select agGridReact={theme} isClearable values={commonOptions} value="option-1" />,
    )
    expect(component.instance().renderValue()).toMatchObject({
      label: "Option 1",
      value: "option-1",
    })
  })

  it("should render the matching object from the options array of strings", () => {
    const options = ["alpha", "bravo", "charlie"]
    const component = shallow(
      <Select agGridReact={theme} isClearable values={options} value="alpha" />,
    )
    expect(component.instance().renderValue()).toMatchObject({
      label: "alpha",
      value: "alpha",
    })
  })

  it("should return a falsy value when given a falsy value", () => {
    const component = shallow(
      <Select agGridReact={theme} isClearable values={commonOptions} value={null} />,
    )
    expect(component.instance().renderValue()).toBeFalsy()
  })

  it("should render top when it is more than 250px from the top", () => {
    const component = shallow(
      <Select agGridReact={theme} isClearable values={commonOptions} value={null} fromTop={251} />,
    )
    expect(component.instance().menuPlacement()).toEqual("top")
  })

  it("should render bottom when it is less than 250px from the top", () => {
    const component = shallow(
      <Select agGridReact={theme} isClearable values={commonOptions} value={null} fromTop={249} />,
    )
    expect(component.instance().menuPlacement()).toEqual("bottom")
  })

  it("should render bottom when it is 250px from the top", () => {
    const component = shallow(
      <Select agGridReact={theme} isClearable values={commonOptions} value={null} fromTop={250} />,
    )
    expect(component.instance().menuPlacement()).toEqual("bottom")
  })

  it("should render bottom when fromTop is not passed", () => {
    const component = shallow(
      <Select agGridReact={theme} isClearable values={commonOptions} value={null} />,
    )
    expect(component.instance().menuPlacement()).toEqual("bottom")
  })
})
