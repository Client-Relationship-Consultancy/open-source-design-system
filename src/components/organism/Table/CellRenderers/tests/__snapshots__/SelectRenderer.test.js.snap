// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select match last snapshot 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  id="SelectEditor"
  isClearable={true}
  isDisabled={false}
  isMulti={false}
  menuIsOpen={false}
  onBlur={[Function]}
  options={
    Array [
      Object {
        "label": "option 1",
        "value": "option 1",
      },
      Object {
        "label": "option 2",
        "value": "option 2",
      },
    ]
  }
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menuList": [Function],
      "multiValue": [Function],
      "multiValueLabel": [Function],
      "multiValueRemove": [Function],
      "option": [Function],
    }
  }
  value={
    Object {
      "label": "option 1",
      "value": "option 1",
    }
  }
/>
`;

exports[`Test handle Change match last snapshot 1`] = `
<Select
  direction="top"
  isClearable={true}
  onChange={[Function]}
  value="option 1"
  values={
    Array [
      "option 1",
      "option 2",
    ]
  }
/>
`;
